### PR TITLE
fix(payment): INT-4063 Avoid holding inventory if a payment intent co…

### PR DESCRIPTION
**Pull request by mauricio-sg**
…nfirmation fails for StripeV3

## What? [INT-4063](https://jira.bigcommerce.com/browse/INT-4063)
Allow reaching the backend when a payment intent confirmation fails. Allowing inventory to be returned to stock immediately.

## Why?
To avoid inventory be held for several minutes due to a failed payment.

## Testing / Proof
Watch the video >>[here](https://drive.google.com/file/d/14KQHRTmkRlvRHCNYJplAfcaLVNzelY9b/view?usp=sharing)<<

## Sibling PR
https://github.com/bigcommerce/bigpay/pull/3674

@bigcommerce/apex-integrations  @bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 1]
```diff
@@ -159,7 +121,24 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
         return Promise.resolve(this._store.getState());
     }
 
-    private async _processAdditionalAction(error: StripeAdditionalActionError, methodId: string, shouldSaveInstrument: boolean): Promise<any> {
+    private _buildPaymentPayload(methodId: string, formattedPayload: any, shouldSetAsDefaultInstrument: boolean): any {
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:124]: Maybe the return type is Payment? since the sign of submitPayment(payment: Payment)

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 2]
```diff
@@ -159,7 +121,24 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
+
+        return { methodId, paymentData };
+    }
+
+    private async _processAdditionalAction(
+        error: StripeAdditionalActionError,
+        methodId: string,
+        shouldSaveInstrument: boolean,
+        shouldSetAsDefaultInstrument: boolean
+    ): Promise<any> {
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:137]: it cannot be Promise<any>, maybe you can find a proper type

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 3]
```diff
@@ -248,27 +223,23 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
             }
 
             if (error) {
-                reject(error);
+                return reject(error);
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:226]: Is this return needed? at the end you have this code wrapped by the Promise, so if you reject a response, then the Promise understand that is a returned response

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 4]
```diff
@@ -248,27 +223,23 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
             }
 
-            if (paymentMethod.method === StripeElementType.CreditCard) {
-                resolve(stripePaymentMethod);
+            if (stripePaymentMethod && paymentMethod.method === StripeElementType.CreditCard) {
+                return resolve(stripePaymentMethod);
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:230]: Same comment as above

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 5]
```diff
@@ -248,27 +223,23 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
             }
 
-            resolve(paymentIntent);
+            if (paymentIntent) {
+                return resolve(paymentIntent);
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:234]: same comment as above about the return

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 6]
```diff
@@ -248,27 +223,23 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
+            }
         });
     }
 
-    private async _confirmVaultedPayment(clientSecret: string): Promise<PaymentIntent> {
-        return new Promise(async (resolve, reject) => {
-            const { error, paymentIntent } = await this._getStripeJs().handleCardAction(clientSecret);
+    private async _handleCustomerAction(clientSecret: string): Promise<PaymentIntent> {
+        const { paymentIntent } = await this._getStripeJs().handleCardAction(clientSecret);
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:240]: Is it not possible that this call answers an error as before?

File: src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts [hunk 7]
```diff
@@ -575,48 +575,6 @@ describe('StripeV3PaymentStrategy', () => {
                     expect(stripeV3JsMock.confirmCardPayment).not.toHaveBeenCalled();
                 });
 
-                it('throws an error when paymentIntent.id is not set properly into payload and is Vaulted', async () => {
```
Comment (clopezh) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts:578]: I see that you remove some tests, and you add new code, why did not you add a tests that covers the new code?

**Review by clopezh — CHANGES_REQUESTED** [review]
Hi @mauricio-sg I have some questions/comments, let me know if you have a question with this

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 8]
```diff
@@ -55,16 +57,15 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
 
     async execute(orderRequest: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
         const { payment, ...order } = orderRequest;
-        let paymentResult: PaymentIntent | StripePaymentMethod;
-        let paymentPayload: any;
+        let formattedPayload: any;
```
Comment (vmparra) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:60]: Any?

Maybe you can at least use something like `let formattedPayload: { }`; But I would rather recommend to create an interface for it.

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts (hunk 1 above)
Comment (vmparra) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:124]: It's better to avoid the use `any` instead you can use `unknown` as it is less permissive

**Review by vmparra — CHANGES_REQUESTED** [review]
Hi @mauricio-sg So far it looks real solid, a few minor comments. Please let me know if you have questions.

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts (hunk 3 above)
Comment (mauricio-sg) [reply src/payment/strategies/stripev3/stripev3-payment-strategy.ts:226]: It's a good practice. This is better explained here: https://stackoverflow.com/questions/32536049/do-i-need-to-return-after-early-resolve-reject/

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts (hunk 6 above)
Comment (mauricio-sg) [reply src/payment/strategies/stripev3/stripev3-payment-strategy.ts:240]: That's the PR's point. I removed on purpose the logic you mention because we need to reach the backend when an error happens in order to return the inventory.

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts (hunk 8 above)
Comment (mauricio-sg) [reply src/payment/strategies/stripev3/stripev3-payment-strategy.ts:60]: If you don't mind, we'll raise a new ticket to address this because `formattedPayload` has a strange structure and needs further investigation and testing. 😬 @cbandam

**Review by clopezh — APPROVED** [review]
LGTM 👍🏻  Thanks to addressed my comments

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts (hunk 8 above)
Comment (clopezh) [reply src/payment/strategies/stripev3/stripev3-payment-strategy.ts]: I think Victor referred in a create an specific type for this scenario, not for all formattedPayloads, maybe something like StripeV3FormattedPayload or some similar of that, Or Do I have a different idea @vmparra  ?

**Review by vmparra — APPROVED** [review]
LGTM 👍🏻

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 9]
```diff
@@ -248,27 +224,23 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
 
             if (error) {
-                reject(error);
+                return reject(error);
             }
 
-            if (paymentMethod.method === StripeElementType.CreditCard) {
-                resolve(stripePaymentMethod);
+            if (stripePaymentMethod && paymentMethod.method === StripeElementType.CreditCard) {
+                return resolve(stripePaymentMethod);
```
Comment (capsula4) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:231]: I just realised this method is `async`, so you don't really need to wrap it inside a Promise and use `resolve/reject`. Async methods automatically wrap the returned value in a promise. If you want to reject a value, you should throw an exception.

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 10]
```diff
@@ -159,7 +122,24 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
+    }
+
+    private async _processAdditionalAction(
+        error: StripeAdditionalActionError,
+        methodId: string,
+        shouldSaveInstrument: boolean,
+        shouldSetAsDefaultInstrument: boolean
+    ): Promise<InternalCheckoutSelectors | never> {
+        if (!(error instanceof RequestError)) {
+            return Promise.reject(error);
```
Comment (capsula4) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:140]: `throw error;` is preferred.

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts [hunk 11]
```diff
@@ -73,79 +75,40 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
...
-                        formattedPayload: {
-                            credit_card_token: {
-                                token: paymentResult.id,
-                            },
-                            confirm: true,
-                        },
-                    },
-                };
-            }
+            // tslint:disable-next-line: variable-name
```
Comment (capsula4) [inline src/payment/strategies/stripev3/stripev3-payment-strategy.ts:78]: why is this required?

**Comment by capsula4** [comment]
Looking good, just some minor syntactic issues.

File: src/payment/strategies/stripev3/stripev3-payment-strategy.ts (hunk 11 above)
Comment (mauricio-sg) [reply src/payment/strategies/stripev3/stripev3-payment-strategy.ts:78]: Because linter complains about `credit_card_number_confirmation`
and `verification_value`. Do you want me to remove it?

**Comment by mauricio-sg** [comment]
@capsula4 ♻️ 🙇

**Review by capsula4 — APPROVED** [review]

**Comment by JoseBobadilla20** [comment]
💚 Tested
QA Details: I ran a regression for Stripe V3 to validate empty payment notification to check that works correctly.
